File: app_mvp/src/main/java/com/example/jqorz/test1/mvp/delegate/MvpInternalDelegate.java
package com.example.jqorz.test1.mvp.delegate;


import com.example.jqorz.test1.mvp.MvpView;
import com.example.jqorz.test1.mvp.presenters.MvpPresenter;

/**
 * Created by jqorz on 2017/8/28.
 * 用于控制MvpPresenter与MvpView:
 */

public class MvpInternalDelegate<P extends MvpPresenter, V extends MvpView> {
    BaseDelegateCallback<P, V> callback;

    public MvpInternalDelegate(BaseDelegateCallback<P, V> callback) {
        this.callback = callback;
    }


    public P createPresenter() {
        P p = callback.getPresenter();
        if (p == null)
            p = callback.createPresenter();
        if (p == null)
            throw new NullPointerException("callback.createPresenter() is null in MvpInternalDelegate");
        return p;
    }

    public void attachView() {
        callback.getPresenter().attachView(callback.getMvpView());
    }

    public void detachView() {
        callback.getPresenter().detachView();
    }
} 

File: app_mvp/src/main/java/com/example/jqorz/mvptest/lib_kotlin/delegate/MvpInternalDelegate.kt
package com.jqorz.tallybook.base.mvp.delegate

import com.jqorz.tallybook.base.mvp.MvpView
import com.jqorz.tallybook.base.mvp.presenters.MvpPresenter

/**
 * 用于控制MvpPresenter与MvpView:
 */
class MvpInternalDelegate<P : MvpPresenter<V>?, V : MvpView>(var callback: BaseDelegateCallback<P, V>) {
    fun createPresenter(): P {
        var p = callback.getPresenter()
        if (p == null) p = callback.createPresenter()
        if (p == null) throw NullPointerException("callback.createPresenter() is null in MvpInternalDelegate")
        return p
    }

    fun attachView() {
        callback.getPresenter()?.attachView(callback.getMvpView())
    }

    fun detachView() {
        callback.getPresenter()?.detachView()
    }
}